File: src/hollywood/storage.py
from __future__ import annotations

import json
from collections.abc import Iterable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

import duckdb
import polars as pl

from .models import (
    ArchivedPayload,
    NormalizedBundle,
    RunStatus,
    SourceDefinition,
    make_stable_id,
)

DDL = (
    """
    create table if not exists source_runs (
        run_id text primary key,
        source_id text,
        source_kind text,
        status text,
        started_at timestamp,
        completed_at timestamp,
        options_json text,
        summary_json text,
        error_text text
    )
    """,
    """
    create table if not exists raw_records (
        raw_record_id text primary key,
        run_id text,
        source_id text,
        source_kind text,
        payload_type text,
        logical_id text,
        source_url text,
        canonical_url text,
        content_type text,
        content_hash text,
        content_path text,
        fetched_at timestamp,
        metadata_json text
    )
    """,
    """
    create table if not exists articles (
        article_id text primary key,
        source_id text,
        canonical_url text,
        url text,
        title text,
        author text,
        published_at timestamp,
        summary text,
        feed_guid text,
        license_class text,
        run_id text,
        metadata_json text
    )
    """,
    """
    create table if not exists article_bodies (
        body_id text primary key,
        article_id text,
        source_id text,
        body_kind text,
        text text,
        raw_record_id text,
        content_hash text,
        license_class text,
        metadata_json text
    )
    """,
    """
    create table if not exists entities (
        entity_id text primary key,
        source_id text,
        external_id text,
        entity_type text,
        name text,
        canonical_name text,
        license_class text,
        metadata_json text
    )
    """,
    """
    create table if not exists entity_aliases (
        entity_alias_id text primary key,
        entity_id text,
        source_id text,
        alias text,
        metadata_json text
    )
    """,
    """
    create table if not exists article_entities (
        article_entity_id text primary key,
        article_id text,
        entity_id text,
        source_id text,
        relation text,
        metadata_json text
    )
    """,
    """
    create table if not exists credits (
        credit_id text primary key,
        source_id text,
        person_entity_id text,
        person_name text,
        title_name text,
        title_external_id text,
        role text,
        billing integer,
        metadata_json text
    )
    """,
)


class HollywoodStorage:
    def __init__(self, db_path: str | Path):
        self.db_path = Path(db_path).resolve()

    def connect(self) -> duckdb.DuckDBPyConnection:
        return duckdb.connect(str(self.db_path))

    def initialize(self) -> None:
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        with self.connect() as conn:
            for statement in DDL:
                conn.execute(statement)

    def start_run(self, source: SourceDefinition, options_json: str) -> str:
        run_id = make_stable_id(source.source_id, options_json, datetime.now(UTC).isoformat())
        with self.connect() as conn:
            conn.execute(
                """
                insert into source_runs (
                    run_id, source_id, source_kind, status, started_at, completed_at, options_json, summary_json, error_text
                ) values (?, ?, ?, ?, now(), null, ?, '{}', null)
                """,
                [
                    run_id,
                    source.source_id,
                    source.kind.value,
                    RunStatus.RUNNING.value,
                    options_json,
                ],
            )
        return run_id

    def finish_run(
        self, run_id: str, status: RunStatus, summary: dict[str, Any], error_text: str | None = None
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                update source_runs
                set status = ?, completed_at = now(), summary_json = ?, error_text = ?
                where run_id = ?
                """,
                [
                    status.value,
                    json.dumps(summary, ensure_ascii=False, sort_keys=True),
                    error_text,
                    run_id,
                ],
            )

    def insert_raw_records(self, run_id: str, archived_payloads: Iterable[ArchivedPayload]) -> None:
        rows = []
        for payload in archived_payloads:
            rows.append(
                {
                    "raw_record_id": payload.raw_record_id,
                    "run_id": run_id,
                    "source_id": payload.source_id,
                    "source_kind": payload.source_kind,
                    "payload_type": payload.payload_type,
                    "logical_id": payload.logical_id,
                    "source_url": payload.source_url,
                    "canonical_url": payload.canonical_url,
                    "content_type": payload.content_type,
                    "content_hash": payload.content_hash,
                    "content_path": payload.content_path,
                    "fetched_at": payload.fetched_at,
                    "metadata_json": payload.metadata_json,
                }
            )
        self._upsert_dicts("raw_records", rows, ("raw_record_id",))

    def apply_bundle(self, bundle: NormalizedBundle) -> None:
        self._upsert_models("articles", bundle.articles, ("article_id",))
        self._upsert_models("article_bodies", bundle.article_bodies, ("body_id",))
        self._upsert_models("entities", bundle.entities, ("entity_id",))
        self._upsert_models("entity_aliases", bundle.entity_aliases, ("entity_alias_id",))
        self._upsert_models("article_entities", bundle.article_entities, ("article_entity_id",))
        self._upsert_models("credits", bundle.credits, ("credit_id",))

    def load_raw_records(
        self, source_id: str | None = None, run_id: str | None = None
    ) -> list[dict[str, Any]]:
        query = "select * from raw_records"
        conditions: list[str] = []
        params: list[Any] = []
        if source_id:
            conditions.append("source_id = ?")
            params.append(source_id)
        if run_id:
            conditions.append("run_id = ?")
            params.append(run_id)
        if conditions:
            query += " where " + " AND ".join(conditions)
        query += " order by fetched_at asc"
        with self.connect() as conn:
            arrow_table = conn.execute(query, params).to_arrow_table()
        return pl.DataFrame(arrow_table).to_dicts()

    def table_count(self, table: str) -> int:
        with self.connect() as conn:
            row = conn.execute(f"select count(*) from {table}").fetchone()
        return int(row[0]) if row is not None else 0

    def export_table(self, table: str, output_dir: Path, file_format: str) -> Path:
        output_dir.mkdir(parents=True, exist_ok=True)
        path = output_dir / f"{table}.{'parquet' if file_format == 'parquet' else 'jsonl'}"
        with self.connect() as conn:
            if file_format == "parquet":
                conn.execute(f"COPY (SELECT * FROM {table}) TO ? (FORMAT PARQUET)", [str(path)])
            else:
                frame = pl.DataFrame(conn.execute(f"select * from {table}").to_arrow_table())
                frame.write_ndjson(path)
        return path

    def export_all(self, output_dir: Path, file_format: str) -> list[Path]:
        return [
            self.export_table(table, output_dir, file_format)
            for table in (
                "source_runs",
                "raw_records",
                "articles",
                "article_bodies",
                "entities",
                "entity_aliases",
                "article_entities",
                "credits",
            )
        ]

    def _upsert_models(self, table: str, rows: Iterable[Any], key_cols: tuple[str, ...]) -> None:
        payload = [row.model_dump(mode="python") for row in rows]
        self._upsert_dicts(table, payload, key_cols)

    def _upsert_dicts(
        self, table: str, rows: list[dict[str, Any]], key_cols: tuple[str, ...]
    ) -> None:
        if not rows:
            return
        # Deduplicate on key columns in case the bundle has the same entity
        # appearing multiple times (e.g., same person credited in multiple titles)
        seen: set[tuple] = set()
        deduped: list[dict[str, Any]] = []
        for row in rows:
            key = tuple(row.get(col) for col in key_cols)
            if key not in seen:
                seen.add(key)
                deduped.append(row)
        frame = pl.DataFrame(deduped, strict=False)
        key_match = " AND ".join(f"t.{column} = s.{column}" for column in key_cols)
        column_list = ", ".join(frame.columns)
        with self.connect() as conn:
            conn.register("_incoming", frame.to_arrow())
            conn.execute(f"DELETE FROM {table} AS t USING _incoming AS s WHERE {key_match}")
            conn.execute(f"INSERT INTO {table} ({column_list}) SELECT {column_list} FROM _incoming")
            conn.unregister("_incoming")

File: api/src/ingest/storage.ts
import { mkdirSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { getDb } from "../db/index.js";
import type { DbRow } from "../db/index.js";
import { makeStableId } from "./models.js";
import type {
  ArchivedPayload,
  ArticleContentRow,
  ArticleEntityRow,
  ArticleRow,
  CreditRow,
  EntityAliasRow,
  EntityRow,
  NormalizedBundle,
  RunStatus,
} from "./models.js";
import type { Candidate } from "./extraction.js";

function now(): string {
  return new Date().toISOString();
}

// ── Run tracking ─────────────────────────────────────────────────────────────

export function startRun(sourceId: string, optionsJson: string): string {
  const runId = makeStableId(sourceId, optionsJson, now());
  getDb()
    .prepare(
      `INSERT INTO runs (id, source_id, run_kind, status, started_at, options_json)
       VALUES (?, ?, 'ingest', ?, ?, ?)`,
    )
    .run(runId, sourceId, "running" satisfies RunStatus, now(), optionsJson);
  return runId;
}

export function finishRun(
  runId: string,
  status: RunStatus,
  summary: Record<string, unknown>,
  errorText?: string,
): void {
  getDb()
    .prepare(
      `UPDATE runs SET status = ?, completed_at = ?, summary_json = ?, error_text = ? WHERE id = ?`,
    )
    .run(status, now(), JSON.stringify(summary), errorText ?? null, runId);
}

export function startRunRaw(runKind: string, metadata: Record<string, unknown>): string {
  const runId = makeStableId(runKind, JSON.stringify(metadata), now());
  getDb()
    .prepare(
      `INSERT INTO runs (id, source_id, run_kind, status, started_at, options_json)
       VALUES (?, 'hollywood', ?, ?, ?, ?)`,
    )
    .run(runId, runKind, "running" satisfies RunStatus, now(), JSON.stringify(metadata));
  return runId;
}

export function insertExtractionRawRecord(
  runId: string,
  sourceId: string,
  contentPath: string,
  contentHash: string,
): string {
  const rawId = makeStableId("extraction_raw", runId, sourceId, contentHash);
  getDb()
    .prepare(
      `INSERT OR IGNORE INTO raw_records
       (id, run_id, source_id, source_kind, payload_type, content_path,
        content_hash, content_type, fetched_at)
       VALUES (?, ?, ?, 'upload', 'text/plain', ?, ?, 'text/plain', ?)`,
    )
    .run(rawId, runId, sourceId, contentPath, contentHash, now());
  return rawId;
}

export function saveExtractionResult(
  runId: string,
  sourceId: string,
  candidate: Candidate,
  modelName: string,
  promptVersion: string,
  rawJson: string,
  rawRecordId?: string,
): void {
  const docId = rawRecordId ?? runId;
  const resultId = makeStableId("extraction", docId, sourceId, candidate.name);
  getDb()
    .prepare(
      `INSERT OR REPLACE INTO extraction_results
       (id, document_id, job_id, schema_version, prompt_version, model_name,
        status, raw_json, result_json, created_at)
       VALUES (?, ?, ?, ?, ?, ?, 'succeeded', ?, ?, ?)`,
    )
    .run(
      resultId,
      docId,
      runId,
      "v1_submission_packet",
      promptVersion,
      modelName,
      rawJson,
      JSON.stringify(candidate),
      now(),
    );
}

export function materializeCandidate(
  candidate: Candidate,
  sourceId = "llm_extraction",
): string {
  const ts = now();
  const db = getDb();
  const entityId = makeStableId("entity", sourceId, candidate.name);

  db.prepare(
    `INSERT OR IGNORE INTO entities
     (id, source_id, entity_type, name, canonical_name, bio, position,
      status, license_class, created_at, updated_at)
     VALUES (?, ?, 'person', ?, ?, ?, ?, 'active', 'public', ?, ?)`,
  ).run(entityId, sourceId, candidate.name, candidate.name.toLowerCase(), candidate.bio, candidate.position ?? "", ts, ts);

  const aliasId = makeStableId("alias", entityId, candidate.name);
  db.prepare(
    `INSERT OR IGNORE INTO entity_aliases (id, entity_id, source_id, alias, created_at)
     VALUES (?, ?, ?, ?, ?)`,
  ).run(aliasId, entityId, sourceId, candidate.name, ts);

  if (candidate.email) {
    const emailId = makeStableId("contact", entityId, candidate.email);
    db.prepare(
      `INSERT OR IGNORE INTO entity_contacts (id, entity_id, source_id, contact_type, contact_value, created_at)
       VALUES (?, ?, ?, 'email', ?, ?)`,
    ).run(emailId, entityId, sourceId, candidate.email, ts);
  }

  if (candidate.phone_number) {
    const phoneId = makeStableId("contact", entityId, candidate.phone_number);
    db.prepare(
      `INSERT OR IGNORE INTO entity_contacts (id, entity_id, source_id, contact_type, contact_value, created_at)
       VALUES (?, ?, ?, 'phone', ?, ?)`,
    ).run(phoneId, entityId, sourceId, candidate.phone_number, ts);
  }

  for (const c of candidate.credits) {
    const titleId = makeStableId("entity", sourceId, c.production);
    db.prepare(
      `INSERT OR IGNORE INTO entities
       (id, source_id, entity_type, name, canonical_name, title_type, status, license_class, created_at, updated_at)
       VALUES (?, ?, 'title', ?, ?, 'tv', 'active', 'public', ?, ?)`,
    ).run(titleId, sourceId, c.production, c.production.toLowerCase(), ts, ts);

    const creditId = makeStableId("credit", entityId, titleId, c.role);
    db.prepare(
      `INSERT OR IGNORE INTO credits (id, person_id, title_id, source_id, role, credit_type, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
    ).run(creditId, entityId, titleId, sourceId, c.role, c.type || "crew", ts);
  }

  for (const org of candidate.organizations) {
    const orgId = makeStableId("entity", sourceId, org.name);
    db.prepare(
      `INSERT OR IGNORE INTO entities
       (id, source_id, entity_type, name, canonical_name, company_type, status, license_class, created_at, updated_at)
       VALUES (?, ?, 'company', ?, ?, ?, 'active', 'public', ?, ?)`,
    ).run(orgId, sourceId, org.name, org.name.toLowerCase(), org.type || "organization", ts, ts);
  }

  for (const tagText of candidate.tags) {
    const norm = tagText.toLowerCase().replace(/ /g, "_");
    const tagId = makeStableId("tag", tagText);
    db.prepare(
      `INSERT OR IGNORE INTO tags (id, tag, normalized_tag, created_at) VALUES (?, ?, ?, ?)`,
    ).run(tagId, tagText, norm, ts);
    const existingTag = db.prepare("SELECT id FROM tags WHERE normalized_tag = ?").get(norm) as { id: string } | undefined;
    const actualTagId = existingTag?.id ?? tagId;
    const taggingId = makeStableId("tagging", entityId, tagText);
    db.prepare(
      `INSERT OR IGNORE INTO entity_taggings (id, tag_id, entity_id, source_id, created_at)
       VALUES (?, ?, ?, ?, ?)`,
    ).run(taggingId, actualTagId, entityId, sourceId, ts);
  }

  for (const link of candidate.links) {
    const linkId = makeStableId("link", entityId, link.url);
    db.prepare(
      `INSERT OR IGNORE INTO entity_links (id, entity_id, source_id, url, link_type, created_at)
       VALUES (?, ?, ?, ?, ?, ?)`,
    ).run(linkId, entityId, sourceId, link.url, link.type, ts);
  }

  for (const rep of candidate.representatives) {
    const repEntityId = makeStableId("entity", sourceId, rep.name);
    db.prepare(
      `INSERT OR IGNORE INTO entities
       (id, source_id, entity_type, name, canonical_name, company_type, status, license_class, created_at, updated_at)
       VALUES (?, ?, 'person', ?, ?, 'agent', 'active', 'public', ?, ?)`,
    ).run(repEntityId, sourceId, rep.name, rep.name.toLowerCase(), ts, ts);

    const repRelId = makeStableId("rep", entityId, repEntityId);
    db.prepare(
      `INSERT OR IGNORE INTO representation
       (id, client_id, rep_id, rep_type, title, email, phone, source_id, trust_state, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'machine_extracted', ?)`,
    ).run(repRelId, entityId, repEntityId, rep.title, rep.title, rep.email ?? "", rep.phone_number ?? "", sourceId, ts);
  }

  return entityId;
}

// ── Raw records ──────────────────────────────────────────────────────────────

export function insertRawRecords(runId: string, archivedPayloads: ArchivedPayload[]): void {
  const db = getDb();
  const stmt = db.prepare(
    `INSERT OR REPLACE INTO raw_records
     (id, run_id, source_id, source_kind, payload_type, content_path, content_hash,
      content_type, source_url, canonical_url, fetched_at, metadata_json)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  );
  for (const p of archivedPayloads) {
    stmt.run(
      p.rawRecordId,
      runId,
      p.sourceId,
      p.sourceKind,
      p.payloadType,
      p.contentPath,
      p.contentHash,
      p.contentType,
      p.sourceUrl ?? null,
      p.canonicalUrl ?? null,
      p.fetchedAt.toISOString(),
      p.metadataJson,
    );
  }
}

export function loadRawRecords(opts: { sourceId?: string; runId?: string } = {}): DbRow[] {
  const conditions: string[] = [];
  const params: unknown[] = [];
  if (opts.sourceId) {
    conditions.push("source_id = ?");
    params.push(opts.sourceId);
  }
  if (opts.runId) {
    conditions.push("run_id = ?");
    params.push(opts.runId);
  }
  let query = "SELECT * FROM raw_records";
  if (conditions.length) query += " WHERE " + conditions.join(" AND ");
  query += " ORDER BY fetched_at ASC";
  return getDb().prepare(query).all(...params) as DbRow[];
}

// ── Normalized bundle ────────────────────────────────────────────────────────

export function applyBundle(bundle: NormalizedBundle): void {
  upsertArticles(bundle.articles);
  upsertArticleContent(bundle.articleContent);
  upsertEntities(bundle.entities);
  upsertEntityAliases(bundle.entityAliases);
  upsertArticleEntities(bundle.articleEntities);
  upsertCredits(bundle.credits);
}

function dedupeById<T extends object>(rows: T[], idKey: keyof T): T[] {
  const seen = new Map<unknown, T>();
  for (const row of rows) seen.set(row[idKey], row);
  return [...seen.values()];
}

function upsertArticles(rows: ArticleRow[]): void {
  const deduped = dedupeById(rows, "articleId");
  if (!deduped.length) return;
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO articles
     (id, source_id, canonical_url, url, title, author, published_at, summary, feed_guid, license_class, run_id, metadata_json)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  );
  for (const r of deduped) {
    stmt.run(
      r.articleId,
      r.sourceId,
      r.canonicalUrl,
      r.url,
      r.title,
      r.author ?? null,
      r.publishedAt?.toISOString() ?? null,
      r.summary ?? null,
      r.feedGuid ?? null,
      r.licenseClass,
      r.runId,
      r.metadataJson,
    );
  }
}

function upsertArticleContent(rows: ArticleContentRow[]): void {
  const deduped = dedupeById(rows, "contentId");
  if (!deduped.length) return;
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO article_content
     (id, article_id, source_id, content_kind, text, raw_record_id, content_hash, license_class, metadata_json)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  );
  for (const r of deduped) {
    stmt.run(r.contentId, r.articleId, r.sourceId, r.contentKind, r.text, r.rawRecordId, r.contentHash, r.licenseClass, r.metadataJson);
  }
}

function upsertEntities(rows: EntityRow[]): void {
  const deduped = dedupeById(rows, "entityId");
  if (!deduped.length) return;
  const ts = now();
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO entities
     (id, source_id, external_id, entity_type, name, canonical_name, status, license_class, metadata_json, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, 'active', ?, ?, ?, ?)`,
  );
  for (const r of deduped) {
    stmt.run(r.entityId, r.sourceId, r.externalId ?? null, r.entityType, r.name, r.canonicalName, r.licenseClass, r.metadataJson, ts, ts);
  }
}

function upsertEntityAliases(rows: EntityAliasRow[]): void {
  const deduped = dedupeById(rows, "entityAliasId");
  if (!deduped.length) return;
  const ts = now();
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO entity_aliases (id, entity_id, source_id, alias, created_at) VALUES (?, ?, ?, ?, ?)`,
  );
  for (const r of deduped) stmt.run(r.entityAliasId, r.entityId, r.sourceId, r.alias, ts);
}

function upsertArticleEntities(rows: ArticleEntityRow[]): void {
  const deduped = dedupeById(rows, "articleEntityId");
  if (!deduped.length) return;
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO article_entities (id, article_id, entity_id, source_id, relation, metadata_json) VALUES (?, ?, ?, ?, ?, ?)`,
  );
  for (const r of deduped) stmt.run(r.articleEntityId, r.articleId, r.entityId, r.sourceId, r.relation, r.metadataJson);
}

function upsertCredits(rows: CreditRow[]): void {
  const deduped = dedupeById(rows, "creditId");
  if (!deduped.length) return;
  const ts = now();
  const stmt = getDb().prepare(
    `INSERT OR REPLACE INTO credits
     (id, person_id, title_id, source_id, role, credit_type, billing, trust_state, metadata_json, created_at)
     VALUES (?, ?, ?, ?, ?, 'cast', ?, 'machine_extracted', ?, ?)`,
  );
  for (const r of deduped) {
    stmt.run(r.creditId, r.personEntityId, r.titleEntityId, r.sourceId, r.role, r.billing ?? null, r.metadataJson, ts);
  }
}

// ── Counts / exports ─────────────────────────────────────────────────────────

export function tableCount(table: string): number {
  const row = getDb().prepare(`SELECT COUNT(*) as count FROM ${table}`).get() as { count: number };
  return row.count;
}

const EXPORTABLE_TABLES = [
  "runs",
  "raw_records",
  "articles",
  "article_content",
  "article_entities",
  "entities",
  "entity_aliases",
  "credits",
] as const;

export function exportTable(table: string, outputDir: string, format: "jsonl" | "parquet"): string {
  if (!EXPORTABLE_TABLES.includes(table as (typeof EXPORTABLE_TABLES)[number])) {
    throw new Error(`Unknown table: ${table}`);
  }
  if (format === "parquet") {
    throw new Error("parquet export is not yet supported in the TypeScript port; use format=jsonl");
  }
  mkdirSync(outputDir, { recursive: true });
  const path = resolve(outputDir, `${table}.jsonl`);
  const rows = getDb().prepare(`SELECT * FROM ${table}`).all() as DbRow[];
  const lines = rows.map((row) => JSON.stringify(row)).join("\n");
  writeFileSync(path, lines ? lines + "\n" : "");
  return path;
}

export function exportAll(outputDir: string, format: "jsonl" | "parquet"): string[] {
  return EXPORTABLE_TABLES.map((table) => exportTable(table, outputDir, format));
}
